server/src/main/java/org/elasticsearch/cluster/ClusterInfoSimulator.java

```diff
@@ -156,17 +185,14 @@ private static long withinRange(long min, long max, long value) {
     }
 
     public ClusterInfo getClusterInfo() {
-        return new ClusterInfo(
-            leastAvailableSpaceUsage,
-            mostAvailableSpaceUsage,
-            shardSizes.toImmutableMap(),
-            shardDataSetSizes,
-            dataPath,
-            Map.of(),
-            estimatedHeapUsages,
-            shardMovementWriteLoadSimulator.simulatedNodeUsageStatsForThreadPools(),
-            allocation.clusterInfo().getShardWriteLoads(),
-            maxHeapSizePerNode
-        );
+        return allocation.clusterInfo()
+            .updateWith(
+                leastAvailableSpaceUsage,
+                mostAvailableSpaceUsage,
+                shardSizes.toImmutableMap(),
+                Map.of(),
+                estimatedHeapUsages,
+                shardMovementWriteLoadSimulator.simulatedNodeUsageStatsForThreadPools()
+            );
```

Comment:
This is the centre piece of the fix. See also commit b78f608ed262a431197b6cb2573f1a5b3326814f for full details.

Note that I will _not_ merge this PR until I get proof from benchmarks.